File: Chat Example/src/main/java/client/ConnectionToChatServerHandler.java
package client;

import chat.ChatProtocols;
import chat.ChatProtocols.Clientbound;
import hoten.serving.ConnectionToServerHandler;
import hoten.serving.JsonMessageBuilder;
import hoten.serving.Message;
import java.io.DataInputStream;
import java.io.IOException;
import java.net.Socket;
import java.util.Map;
import static chat.ChatProtocols.Serverbound.*;

public class ConnectionToChatServerHandler extends ConnectionToServerHandler {

    private final Chat _chat;

    public ConnectionToChatServerHandler(Chat chat, Socket socket) throws IOException {
        super(socket, new ChatProtocols());
        _chat = chat;
    }

    public void quit() {
        Message message = new JsonMessageBuilder()
                .protocol(outbound(LogOff))
                .build();
        send(message);
    }

    public void sendMessage(String msg) {
        Message message = new JsonMessageBuilder()
                .protocol(outbound(ChatMessage))
                .set("msg", msg)
                .build();
        send(message);
    }

    public void sendWhisper(String to, String msg) {
        Message message = new JsonMessageBuilder()
                .protocol(outbound(PrivateMessage))
                .set("to", to)
                .set("msg", msg)
                .build();
        send(message);
    }

    @Override
    protected void onConnectionSettled() {
        _chat.startChat(this);
    }

    @Override
    protected void handleData(int type, Map data) throws IOException {
        switch (Clientbound.values()[type]) {
            case PeerJoin:
                _chat.announceNewUser((String) data.get("username"));
                break;
            case ChatMessage:
                _chat.global((String) data.get("from"), (String) data.get("msg"));
                break;
            case PeerDisconnect:
                _chat.announceDisconnect((String) data.get("username"));
                break;
            case PrivateMessage:
                _chat.whisper((String) data.get("from"), (String) data.get("msg"));
                break;
            case Print:
                _chat.announce((String) data.get("msg"));
                break;
        }
    }

    @Override
    protected void handleData(int type, DataInputStream data) {
    }

    void sendUsername(String username) {
        Message message = new JsonMessageBuilder()
                .protocol(outbound(SetUsername))
                .set("username", username)
                .build();
        send(message);
    }
}

File: Chat Example/src/main/csharp/ChatClient/ChatClient/ConnectionToChatServerHandler.cs
using Serving;
using System;
using System.Collections.Generic;
using System.IO;

namespace ChatClient
{
    class ConnectionToChatServerHandler : ConnectionToServerHandler
    {
        private Chat _chat;

        public ConnectionToChatServerHandler(Chat chat, String host, int port)
            : base(host, port, new ChatProtocols(), BoundDest.SERVER)
        {
            _chat = chat;
        }

        protected override void OnConnectionSettled()
        {
            _chat.StartChat(this);
        }

        protected override void HandleData(int type, Dictionary<String, String> data)
        {
            switch ((ChatProtocols.Clientbound)type)
            {
                case ChatProtocols.Clientbound.PeerJoin:
                    _chat.AnnounceNewUser(data["username"]);
                    break;
                case ChatProtocols.Clientbound.ChatMessage:
                    _chat.Global(data["from"], data["msg"]);
                    break;
                case ChatProtocols.Clientbound.PeerDisconnect:
                    _chat.AnnounceDisconnect(data["username"]);
                    break;
                case ChatProtocols.Clientbound.PrivateMessage:
                    _chat.Whisper(data["from"], data["msg"]);
                    break;
                case ChatProtocols.Clientbound.Print:
                    _chat.Announce(data["msg"]);
                    break;
            }
        }

        protected override void HandleData(int type, BinaryReader data) { }

        public void SendUsername(String username)
        {
            Message message = new JsonMessageBuilder()
                    .Protocol(Outbound(ChatProtocols.Serverbound.SetUsername))
                    .Set("username", username)
                    .Build();
            Send(message);
        }

        public void Quit()
        {
            Message message = new JsonMessageBuilder()
                    .Protocol(Outbound(ChatProtocols.Serverbound.LogOff))
                    .Build();
            Send(message);
        }

        public void SendMessage(String msg)
        {
            Message message = new JsonMessageBuilder()
                    .Protocol(Outbound(ChatProtocols.Serverbound.ChatMessage))
                    .Set("msg", msg)
                    .Build();
            Send(message);
        }

        public void SendWhisper(String to, String msg)
        {
            Message message = new JsonMessageBuilder()
                    .Protocol(Outbound(ChatProtocols.Serverbound.PrivateMessage))
                    .Set("to", to)
                    .Set("msg", msg)
                    .Build();
            Send(message);
        }
    }
}
